velox/type/TimestampConversion.cpp

```diff
@@ -812,6 +814,83 @@ Status parserError(const char* str, size_t len) {
       std::string(str, len));
 }
 
+inline bool startsWith(std::string_view str, const char* prefix) {
+  return str.rfind(prefix, 0) == 0;
+}
+
+bool parseSparkTzWithPrefix(
+    std::string& timeZoneName,
+    Timestamp& ts,
+    size_t start) {
+  VELOX_DCHECK_GT(timeZoneName.length(), start + 2);
+  auto sign = timeZoneName[start];
+  auto offset = timeZoneName.substr(start + 1);
+  std::string hm, h, m, s;
+  RE2 hhPattern(R"((\d?\d))");
+  RE2 hhmmPattern(R"((\d\d):?(\d\d))");
+  RE2 hhmmssPattern(R"((\d\d:\d\d):(\d\d))");
+  RE2 hhmmssCompactPattern(R"((\d\d)(\d\d)(\d\d))");
+  if (RE2::FullMatch(offset, hhPattern, &hm)) {
+    hm += ":00";
+  } else if (RE2::FullMatch(offset, hhmmPattern, &h, &m)) {
+    hm = h + ":" + m;
+  } else if (RE2::FullMatch(offset, hhmmssPattern, &hm, &s)) {
+    // hm and second already captured
+  } else if (RE2::FullMatch(offset, hhmmssCompactPattern, &h, &m, &s)) {
+    hm = h + ":" + m;
+  } else {
+    return false;
+  }
+  int32_t sec = 0;
+  size_t pos = 0;
+  if (!s.empty() && !parseDoubleDigit(s.c_str(), s.length(), pos, sec)) {
+    return false;
+  }
+  if (sec < 0 || sec > 60) {
+    return false;
+  }
+  if (sec != 0) {
+    // Seconds timezone offset is not support, need to be applied to timestamp.
+    if (sign == '+') {
+      if (ts.getSeconds() - sec < Timestamp::kMinSeconds) {
+        return false;
+      }
+      ts = Timestamp(ts.getSeconds() - sec, ts.getNanos());
+    } else if (sign == '-') {
+      if (ts.getSeconds() + sec > Timestamp::kMaxSeconds) {
+        return false;
+      }
+      ts = Timestamp(ts.getSeconds() + sec, ts.getNanos());
+    } else {
+      return false;
+    }
+  }
+  timeZoneName = sign + hm;
+  return true;
+}
+
+bool normalizeSparkTimezone(std::string& timeZoneName, Timestamp& ts) {
+  // Spark support an id with one of the prefixes UTC+, UTC-, GMT+, GMT-, UT+ or
+  // UT-, and a suffix in the formats: h[h], hh[:]mm, hh:mm:ss, hhmmss
```

Comment:
does Spark accept timezone offsets containing seconds? AFAIK, IANA's definition is only (+/-)00:00 (hours and minutes). There are no use cases for timezone conversions with second granularity.

Reply:
Yes, Spark's timezone support includes the format `UTC+hh:mm:ss`. Therefore, I can only add the seconds offset to the Timestamp result.

Reply:
That is odd. What does Spark do if you have seconds on that definition? Does it add the seconds to the timezone conversion process?

Reply:
Yes, Spark will add the seconds to the timezone conversion process. For example:
```
spark-sql (default)> select cast('2015-03-07 19:33:33 UTC-00:00:01' as timestamp);
2015-03-08 03:33:34
Time taken: 0.044 seconds, Fetched 1 row(s)
```